apps/game_backend/lib/game_backend/users.ex

```diff
@@ -533,4 +573,132 @@ defmodule GameBackend.Users do
+
+        case any_failed_quest? do
+          {:error, changeset} ->
+            {:error, changeset}
+
+          _ ->
+            {:ok, :quests_generated}
+        end
+    end)
+    |> Multi.run(:user, fn _, %{insert_user: user} ->
+      get_user_by_id_and_game_id(user.id, curse_id)
+    end)
+    |> Repo.transaction()
+  end
+
+  def get_user_and_maybe_insert_daily_quests(user_id) do
+    curse_id = GameBackend.Utils.get_game_id(:curse_of_mirra)
+
+    Multi.new()
+    |> Multi.run(:get_user, fn _, _ ->
+      get_user_by_id_and_game_id(user_id, curse_id)
+    end)
+    |> Multi.run(:maybe_generate_quests, fn
+      _, %{get_user: user} ->
+        today = Date.utc_today()
+
+        should_generate_quests? =
+          user.last_daily_quest_generation_at &&
+            Date.compare(NaiveDateTime.to_date(user.last_daily_quest_generation_at), today) == :lt
+
+        if should_generate_quests? do
+          quests_results = generate_daily_quests_for_user(user)
+          any_failed_quest = Enum.find(quests_results, fn {result, _quest} -> result == :error end)
+
+          user_changeset =
+            GameBackend.Users.User.changeset(user, %{last_daily_quest_generation_at: NaiveDateTime.utc_now()})
+
+          user_update = Repo.update(user_changeset)
+
+          case {any_failed_quest, user_update} do
+            {nil, {:ok, user}} ->
+              {:ok, {quests_results, user}}
+
+            {{:error, changeset}, _user_update} ->
+              {:error, changeset}
+
+            {_, {:error, user_changeset}} ->
+              {:error, user_changeset}
+          end
+        else
+          {:ok, :not_needed}
+        end
+    end)
+    |> Multi.run(:user, fn _, %{get_user: user} ->
+      get_user_by_id_and_game_id(user.id, curse_id)
+    end)
+    |> Repo.transaction()
+  end
+
+  defp generate_daily_quests_for_user(user) do
+    available_quests =
+      Quests.get_user_missing_quests_by_type(user.id, "daily")
+      |> Enum.shuffle()
+
+    {active_quests_params, remaining_quests} = Enum.split(available_quests, 3)
+
+    {inactive_quests_params, _remaining_quests} = Enum.split(remaining_quests, 3)
+
+    active_quests =
+      Enum.map(active_quests_params, fn
+        quest_params ->
+          attrs = %{
+            user_id: user.id,
+            quest_id: quest_params.id,
+            status: "available",
+            activated_at: NaiveDateTime.utc_now()
+          }
+
+          changeset = UserQuest.changeset(%UserQuest{}, attrs)
+
+          Repo.insert(changeset)
+      end)
+
+    inactive_quests =
+      Enum.map(inactive_quests_params, fn
+        quest_params ->
+          attrs = %{
+            user_id: user.id,
+            quest_id: quest_params.id,
+            status: "available"
+          }
+
+          changeset = UserQuest.changeset(%UserQuest{}, attrs)
+
+          Repo.insert(changeset)
+      end)
+
+    active_quests ++ inactive_quests
```

Comment:
Is there any reason to return every transaction result even if some of their insertions failed?
If we only want them all to succeed we should implement a multi then return either {:ok, list_of_quests} or {:error, _changesets}

Reply:
agreed!